### PR TITLE
Added support for babel config under package.json/ava key (Fixes #448)

### LINKED ISSUE #448: A possible way to provide customizable babel configs using `.babelrc`/`env`
Continued from:
https://github.com/sindresorhus/ava/pull/398#issuecomment-172386983

My proposal is that we allow the following in `.babelrc`:

```js
{
  "presets": ["es2015"], // base config
  "sourceMaps": "inline",
  "env": {
    "ava" : {  // if `env/ava` is present, we disable our custom config.
      "presets": ["stage-0"],
      "plugins": [] //maybe we automatically append power-assert for them?
    },
    "production" : {
     "sourceMaps": false
    }
  }
}
```

There are a couple questions that need to be answered in order to proceed. Specifically, can we change which `env` Babel uses on a per file basis? It might be as easy as setting `process.env.BABEL_ENV` before each call to `babel.transform(..)`

### PR DESCRIPTION
**Pull request by spudly**
Here's an initial commit that allows package owners to specify a babel config for tests under the package.json/ava key.

This conforms to the proposal in the following comment: https://github.com/sindresorhus/ava/issues/448#issuecomment-187694472

I chose not to implement an extends key discussed in the thread because 1) it adds to the complexity, 2) I personally don't need it, and 3) how to solve that request is still somewhat unresolved. If needed, someone can add that later in another pull request.

Let me know what you think. I'm a little unsure about whether my method for obtaining babelConfig is the best one.

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -97,7 +97,8 @@ var api = new Api(cli.input.length ? cli.input : arrify(conf.files), {
 	failFast: cli.flags.failFast,
 	serial: cli.flags.serial,
 	require: arrify(cli.flags.require),
-	cacheEnabled: cli.flags.cache !== false
+	cacheEnabled: cli.flags.cache !== false,
+  babelConfig: conf.babel,
```
Comment (sindresorhus) [inline cli.js]: Wrong indentation here.

**Comment by sindresorhus** [comment]
Good start :) This will need tests and docs too.

**Comment by spudly** [comment]
Fixed indentation, added tests, and added docs.

File: readme.md [hunk 2]
```diff
@@ -363,6 +364,41 @@ test(t => {
...
+			"react"
+		]
+	},
+	"ava": {
+		"babel": "inherit",
+	},
+}
+```
+
+If you do not specify a "babel" key in your ava configuration, or if you set it to `"default"`, AVA will transpile the test files with AVA's default babel configuration.
```
Comment (kentcdodds) [inline readme.md]: Is this default configuration defined anywhere else in the docs? If not, it should be. Then this should link to that.

**Comment by vdemedes** [comment]
Hey @spudly, could you rebase from master to resolve a conflict? Thanks for the great work!

**Comment by spudly** [comment]
@vdemedes

It's been rebased.

**Comment by vdemedes** [comment]
Great, I think it's good to go!

@sindresorhus @novemberborn @jamestalmage ?

File: lib/caching-precompiler.js [hunk 3]
```diff
@@ -30,15 +31,26 @@ CachingPrecompiler.prototype._factory = function (cacheDir) {
...
 			sourceMaps: true,
 			ast: false,
-			babelrc: false,
-			inputSourceMap: sourceMap && sourceMap.toObject()
+			inputSourceMap: sourceMap && sourceMap.toObject(),
+			babelrc: false
 		};
+
+		if (!babelConfig || babelConfig === 'default') {
+			return objectAssign({}, baseOptions, {
```
Comment (novemberborn) [inline lib/caching-precompiler.js]: No need for the empty object here, `baseOptions` is already fresh so it's safe to mix other properties into it.

File: readme.md [hunk 4]
```diff
@@ -367,6 +368,41 @@ AVA comes with builtin support for ES2015 through [Babel 6](https://babeljs.io).
...
+					"es2015",
+					"stage-0",
+					"react"
+			 ]
+		 }
+	},
+}
+```
+
+In addition to specifying a custom babel config, you can also use the special `inherit` keyword. When you do this, AVA will allow tests to be transpiled using the configuration defined in your .babelrc file or in package.json/babel. This way, your test files will be transpiled using the same options as your source files, but you won't have to define the options twice.
```
Comment (novemberborn) [inline readme.md]: `"inherit"` to be consistent with `"default"` below?

File: readme.md [hunk 5]
```diff
@@ -367,6 +368,41 @@ AVA comes with builtin support for ES2015 through [Babel 6](https://babeljs.io).
 
 AVA includes typings for TypeScript. You have to setup transpilation yourself. When you set `module` to `commonjs` in your `tsconfig.json` file, TypeScript will automatically find the type definitions for AVA. You should set `target` to `es2015` to use Promises and async functions.
 
+If you want to customize the babel transpiler for test files, you can do so by adding a `"babel"` key to the `"ava"` section in your package.json file.
```
Comment (novemberborn) [inline readme.md]: Some nits:

* This needs a heading.
* The README isn't entirely consistent but I'd prefer `package.json` (with the backticks). Same for `.babelrc`.
* It's named "`ava` section" (without quotes) earlier: https://github.com/spudly/ava/blob/f928959b9ad106a70fc3091a462ff3d96a60f469/readme.md#configuration
* Could you capitalize Babel throughout?

**Comment by novemberborn** [comment]
Haven't looked at the tests yet, sorry. Been staring at AVA issues all day :)

Could you follow up on https://github.com/sindresorhus/ava/issues/448#issuecomment-190286370?

Looking really good though, nice work.

File: lib/caching-precompiler.js [hunk 6]
```diff
@@ -30,15 +31,26 @@ CachingPrecompiler.prototype._factory = function (cacheDir) {
...
+				presets: [presetStage2, presetES2015],
+				plugins: [powerAssert, transformRuntime]
+			});
+		}
+
+		if (babelConfig === 'inherit') {
+			return objectAssign({}, baseOptions, {babelrc: true});
```
Comment (jamestalmage) [inline lib/caching-precompiler.js]: As discussed, we will want to ensure `powerAssert` is added as a plugin regardless.

File: readme.md (hunk 2 above)
Comment (jamestalmage) [inline readme.md]: :+1: I don't believe it is defined anywhere, but it should be.

File: lib/caching-precompiler.js [hunk 7]
```diff
@@ -30,15 +31,26 @@ CachingPrecompiler.prototype._factory = function (cacheDir) {
...
-			presets: [presetStage2, presetES2015],
-			plugins: [powerAssert, transformRuntime],
+		var baseOptions = {
 			filename: filename,
 			sourceMaps: true,
 			ast: false,
-			babelrc: false,
-			inputSourceMap: sourceMap && sourceMap.toObject()
+			inputSourceMap: sourceMap && sourceMap.toObject(),
+			babelrc: false
```
Comment (novemberborn) [inline lib/caching-precompiler.js]: Unnecessary diff churn

File: test/caching-precompiler.js [hunk 8]
```diff
@@ -18,9 +20,14 @@ function endsWithMap(filename) {
 	return /\.js$/.test(filename);
 }
 
+test('before', function (t) {
+	sinon.spy(babel, 'transform');
```
Comment (novemberborn) [inline test/caching-precompiler.js]: AVA's tests are run using [`tap`](http://www.node-tap.org/) which runs each test file in its own process. You don't have to worry about modifying the `babel-core` module, or about cleaning it up afterwards. Just add the spy in the module itself, outside of a test.

**Comment by novemberborn** [comment]
The tests exercise `caching-compiler`, which is good. I'd like to see some integration tests in `test/cli`, though that does not have to hold up this PR. Happy to undertake that myself or help you @spudly if you're interested. It might be a bit tricky to verify the differing Babel behaviors.

---

Right now `package.json/ava/babel` allows any Babel option to be changed. To avoid footguns though I think `filename`, `sourceMaps`, `ast` and `inputSourceMap` should not be configurable.

---

@spudly thanks for the hard work on this. It's solving so many of our issues!

**Comment by spudly** [comment]
Pushed a new commit & rebased. I believe I've addressed all of the concerns that were discussed, with the exception of the `babel: false` option, which I believe should be addressed in a separate pull request.

**Comment by spudly** [comment]
Can anyone help explain why these 44 tests are failing?

https://travis-ci.org/sindresorhus/ava/jobs/112878667

They're unrelated to my changes - the master branch of AVA fails the same tests. These tests were passing yesterday, but today they fail. Perhaps one of the dependencies had a release that broke with semver?

**Comment by wyze** [comment]
Might be related to babel 6.6 [release](https://github.com/babel/babel/releases/tag/v6.6.0) yesterday.

I noticed that your previous build used 6.5.2 while this one used 6.6.

File: test/caching-precompiler.js [hunk 9]
```diff
@@ -18,9 +20,11 @@ function endsWithMap(filename) {
 	return /\.js$/.test(filename);
 }
 
+sinon.spy(babel, 'transform');
+
 test('creation with new', function (t) {
 	var tempDir = uniqueTempDir();
-	var precompiler = new CachingPrecompiler(tempDir);
+	var precompiler = new CachingPrecompiler(tempDir, null);
```
Comment (novemberborn) [inline test/caching-precompiler.js]: Should pass `'default'` here (and below).

**Comment by MoOx** [comment]
Failures may be similar to https://github.com/MoOx/statinamic/issues/229 (or may not)

**Comment by novemberborn** [comment]
@spudly did a fresh checkout and install, `master` still passes but this branch does not.

**Comment by novemberborn** [comment]
@spudly ah! The failing tests use `api.js` directly, without setting the `babelConfig` option. Now I feel bad :cry: 

I still think using the default config in `cli.js` is the right move, but maybe add the `!babelConfig || babelConfig === 'default'` clause back in `caching-precompiler.js`. Quite bothersome passing it in the 44 tests that instantiate `Api`.

**Comment by spudly** [comment]
My bad, I must have made a mistake when I checked out master and ran the tests again yesterday. Did it again this morning and the tests are passing. I'll fix this very soon.

**Comment by spudly** [comment]
I'm stumped guys. If I do the following, all tests pass as expected.

```bash
git clone https://github.com/sindresorhus/ava.git
cd ava
git reset --hard 58e1c561296a409a9557ceb1c7684685404c4278
npm install
npm test
```

But if I do this, 44 tests fail:

```bash
git clone https://github.com/spudly/ava.git
cd ava
git reset --hard 58e1c561296a409a9557ceb1c7684685404c4278
npm install
npm test
```

Both repos are being reset to the same initial commit (the one before any of my changes), so there should be absolutely no difference between the two, right?

**Comment by kentcdodds** [comment]
I'm running both of those series of commands now. I'll let you know what I get

**Comment by kentcdodds** [comment]
Everything passed for me for both of those. Sounds like an environment issue...

**Comment by kentcdodds** [comment]
But yes, this fails 44 tests for me:

```
git clone https://github.com/spudly/ava.git
cd ava
npm install
npm test
```

**Comment by spudly** [comment]
my current theory is that I'm getting weird test results because of nyc cache. testing that theory now...

**Comment by MoOx** [comment]
Maybe try BABEL_DISABLE_CACHE=1.

**Comment by spudly** [comment]
@moOx,

I didn't know about that flag, but the following just worked for me:

1. `git clone`...
2. `cd ava`
4. `npm install`
3. [make edits]
5. `npm test`

if before step 3 I ran `npm test`, the test at step 5 would fail with 44 errors even though the source code was fixed.

Not sure at this point if it was nyc cache or babel cache but it was definitely caching the source files somewhere and screwing everything up.

**Comment by novemberborn** [comment]
@spudly,

> if before step 3 I ran npm test, the test at step 5 would fail with 44 errors even though the source code was fixed.

Do you have `xo`, `nyc` and/or `tap` installed globally? Else `npm test` wouldn't even be able to cache anything. I wonder if there's some local/global mismatch, presumably in `nyc`. Which `nyc` version do you have globally?

---

Is this ready for another look?

**Comment by spudly** [comment]
@novemberborn, I don't have any of those modules installed globally.

Yes, the pull request is ready.

**Comment by novemberborn** [comment]
> I don't have any of those modules installed globally.

Then I'm stumped, sorry :disappointed:

File: test/caching-precompiler.js [hunk 10]
```diff
@@ -48,3 +52,64 @@ test('adds files and source maps to the cache directory as needed', function (t)
...
+	t.true(babel.transform.calledOnce);
+	var options = babel.transform.firstCall.args[1];
+
+	t.true('filename' in options);
+	t.true(options.sourceMaps);
+	t.false(options.ast);
+	t.true('inputSourceMap' in options);
+	t.false(options.babelrc);
+	t.deepEqual(options.presets, ['stage-2', 'es2015']);
+	t.equal(options.plugins.length, 2);
```
Comment (novemberborn) [inline test/caching-precompiler.js]: Could you add a configured plugin on line 98, and then verify not the `plugins` length but the expected values? Would be good to test that the configured plugins are preserved and the power-assert and runtime transform plugins are appended.

Could then change the `plugins` test in the default case to match for those plugins too. Similarly the inherit case should check the plugins were added.

**Comment by novemberborn** [comment]
@spudly did I mention I'm pretty excited about this landing? I'm pretty excited about this landing :)

Just one suggestion regarding testing the plugins.

Oh, and to confirm, Babel merges plugin definitions when you set `babelrc` to `true`?

File: test/caching-precompiler.js (hunk 10 above)
Comment (spudly) [inline test/caching-precompiler.js]: I can certainly add a fake plugin, which would increase the expected plugins.length to 3. I can also verify that transformRuntime is added because I can just require() it in to get a reference to it and use === for comparison.

But I'm not sure how I can verify that powerAssert was added because it is instantiated inside of the `_factory` function, so I don't have a reference to it. That's why I only verified that lenth === 2.

Is there another way I can validate that the powerAssert plugin was added?

Comment (jamestalmage) [inline test/caching-precompiler.js]: > Is there another way I can validate that the powerAssert plugin was added?

Run a test with a bad assertion that generates powerAssert output. There is an existing test that already does that using the default config. You would just need to create a similar one that does it using custom config.

**Comment by jamestalmage** [comment]
> @spudly did I mention I'm pretty excited about this landing? I'm pretty excited about this landing :)

Agreed. This should be a high priority merge. Lots of people want this feature.

**Comment by kentcdodds** [comment]
> Lots of people want this feature.

[Anyone wanting to use JSX](https://github.com/kentcdodds/react-ava-workshop/blob/start/INSTRUCTIONS.md#ava) in their test files is still using `ava@0.9.2` until this lands. So thank you so much everyone for working on it and making sure that the API is awesome! :tada:

**Comment by jamestalmage** [comment]
> Anyone wanting to use JSX in their test files is still using ava@0.9.2 until this lands

Sorry about that. When we made the decision to ignore `.babelrc`, it wasn't supposed to take this long to get it back in.

**Comment by kentcdodds** [comment]
Heh, I know that. That's open source. If I really really needed to get this in, I could have taken a more proactive approach to making that happen. #priorities ;-) I'm just glad that there's a shared priority here!

File: test/caching-precompiler.js (hunk 10 above)
Comment (novemberborn) [inline test/caching-precompiler.js]: Try this, after you instantiate `new CachingPrecompiler()`

```js
var powerAssert = function () {};
sinon.stub(precompiler, '_createEspowerPlugin').returns(powerAssert);
```

And then compare against the `powerAssert` variable inside your test.

**Comment by spudly** [comment]
Improved unit tests to check that powerAssert and transformRuntime get appended properly.

Ready for evaluation again.

> Oh, and to confirm, Babel merges plugin definitions when you set babelrc to true?

Yes, I believe that is the case.

**Comment by novemberborn** [comment]
LGTM! Great work @spudly.

**Comment by jamestalmage** [comment]
>> Oh, and to confirm, Babel merges plugin definitions when you set babelrc to true?

> Yes, I believe that is the case.

Should we have an integration test confirming that?

Otherwise, lets ship this.

File: readme.md [hunk 11]
```diff
@@ -367,6 +368,61 @@ AVA comes with builtin support for ES2015 through [Babel 6](https://babeljs.io).
...
+			"react"
+		]
+	},
+	"ava": {
+		"babel": "inherit",
+	},
+}
+```
+
+Note: When configuring Babel for tests manually, the espower and transform-runtime plugins will be
```
Comment (jamestalmage) [inline readme.md]: Why `transform-runtime`?

**Comment by kentcdodds** [comment]
FWIW, I just did this:

```
$ npm i spudly/ava#461c781fcaabe29eb29464c3d98f9e19b64a8af2
$ npm test
```

And all tests pass. Verified I had the right version

```
$ node_modules/.bin/ava --version
```

And got `0.12.0` which I think is correct.

The thing that surprised me is that I didn't need to do any configuration to make my tests work with JSX. I was under the impression I had to explicitly add this config:

```javascript
"ava": {
  "babel": "inherit"
}
```

But it looks like my tests were transpiled with my default `.babelrc` anyway. Is that correct?

Either way, it's awesome that I'll be able to use the latest AVA soon! :D

**Comment by jamestalmage** [comment]
> I was under the impression I had to explicitly add this config:

Yes - you should definitely be required to explicitly configure the "inherit" property. 

@spudly - when you debug that, please be sure to add a few integration tests to verify correct behavior. Specifically, that `.babelrc` is only consulted for `inherit`.

**Comment by jamestalmage** [comment]
@spudly - see https://github.com/sindresorhus/ava/commit/1b01f16d9bcd6f5d62e3becb8d60e457e7ac9690

You should be able to merge the `babel-plugin-for-integration-tests` branch into this PR without conflict.

To test out what the babel plugin does:

```sh
$ node_modules/.bin/babel --plugins ./babel-plugin-test-doubler test/fixture/fail-fast.js
```

It should be pretty easy to create integration tests from there using the API. (You should get double the number of tests when the plugin is used).

File: readme.md (hunk 11 above)
Comment (spudly) [inline readme.md]: If we don't include transform-runtime, users will have to `import 'babel-polyfill'` at the top of each test file. This has not changed in the pull request. It was like this before. This is only a documentation change.

**Comment by spudly** [comment]
Apparently when ava compiles a file it caches the result (hence the name of the file, `cachingPrecompiler.js`). This cache is stored in `node_modules/.cache/ava`.

I'm pretty sure this is the issue I was having during a couple of days ago. Simply removing ./node_modules/.cache probably would have solved my problem.

@jamestalmage, is there an easy way to programmatically clear AVA's compile cache (or disable it)? I need to be able to do this so I can isolate the unit tests from each other. The issue is that if I reuse a test fixture (using `test/fixtures/es2015.js` right now), it will compile with one babel config and cache the result. Then, later, when another test is run with a different babel config, the test fails because it uses a cached version.

**Comment by novemberborn** [comment]
> I'm pretty sure this is the issue I was having during a couple of days ago. Simply removing ./node_modules/.cache probably would have solved my problem.

Oh of course!

@spudly you can use the `--no-cache` flag to disable the cache.

**Comment by spudly** [comment]
@kentcdodds, still looking at your issue, but is it possible that your issue is related to the compiler cache as well? I'm thinking maybe you added the react to the list of plugins, ran the compiler (cache files get created here), then removed it and expected it to fail but it worked anyway because it used the cached version the second time. Can you remove `node_modules/.cache` and try again?

**Comment by novemberborn** [comment]
@spudly we'll have to include the `babelConfig` option in the salt for the caching transform, see <https://github.com/sindresorhus/ava/blob/e697629a77b511e069eb0a2991c8ee72cd09034c/lib/caching-precompiler.js#L68>.

Also there's a merge conflict now :(

**Comment by spudly** [comment]
yep, fixing the merge conflict...

**Comment by kentcdodds** [comment]
Success! It probably was the cache. Here's a reproducible command to show old/new behavior.

```
git clone https://github.com/kentcdodds/react-ava-workshop.git && cd react-ava-workshop && npm i && npm i spudly/ava#461c781fcaabe29eb29464c3d98f9e19b64a8af2 && npm t
```

You'll get a parser error.

Now, to the `package.json` add:

```js
{
  "ava": {
    "babel": "inherit"
  }
}
```

And rerun `npm t` and you'll get 10 passing tests :+1:

**Comment by spudly** [comment]
Added babelConfig to the salt and re-enabled caching in unit tests. All working.

Should be ready to merge now, right? (assuming tests pass in travis && appveyor)

**Comment by kentcdodds** [comment]
That's me, watching the the status of the AppVeyor build...

**Comment by kentcdodds** [comment]
Me now that it's all passing!

**Comment by spudly** [comment]
only took 22 minutes... is there a reason we're running tests in both appveyor and travis? Seems like one is enough

**Comment by kentcdodds** [comment]
> is there a reason we're running tests in both appveyor and travis? Seems like one is enough

Gotta make sure windows works (right?)

**Comment by kentcdodds** [comment]
> only took 22 minutes...

Yeah, it only took that to actually run, but it took much longer to start. Believe me, I've been watching :-)

**Comment by spudly** [comment]
> t.threw??

This gives me a better error if the test fails. If the promise in that test is rejected, it will call `t.threw(error)` and the tests will fail with that error. If I didn't include `t.threw` to handle the rejection, the reason for tests failure would just be that there was a missing test, which is not enough information for debugging. This should probably be added to a lot of the other tests as well.

**Comment by novemberborn** [comment]
> If I didn't include t.threw to handle the rejection, the reason for tests failure would just be that there was a missing test, which is not enough information for debugging. This should probably be added to a lot of the other tests as well.

AVA now requires a version of `tap` which supports promises. You should be able to return the promise and `tap` will report the failure.

File: package.json [hunk 12]
```diff
@@ -131,6 +131,7 @@
     "update-notifier": "^0.6.0"
   },
   "devDependencies": {
+    "babel-cli": "^6.6.4",
```
Comment (novemberborn) [inline package.json]: Why this dependency?

**Comment by novemberborn** [comment]
@spudly could you return the `api.run()` promise and remove the `t.threw` bit? And just a question about the `babel-cli` dependency. LGTM otherwise!

---

When squashing would be good to keep @jamestalmage's commit separate.

**Comment by novemberborn** [comment]
Please make sure to rebase on `master` to [fix a linting issue](https://github.com/sindresorhus/ava/pull/601) and run the latest XO configuration against your PR. Thanks!

**Comment by spudly** [comment]
Will do. I'll make these changes either tonight or Monday morning.

**Comment by spudly** [comment]
Is there a reason we have a `return` statement on [cli.js:13](https://github.com/sindresorhus/ava/blob/master/cli.js#L13)? It's causing lint errors with the latest XO config.

**Comment by spudly** [comment]
Nevermind. running npm install fixed that issue. Still curious though, why are those return statements there?

**Comment by spudly** [comment]
made requested changes and rebased again.

**Comment by novemberborn** [comment]
> Is there a reason we have return statements on cli.js:13 and cli.js:96?

They're shorthand for exiting the CLI.

**Comment by novemberborn** [comment]
@spudly what's the reason for adding the `babel-cli` dependency in `package.json`?

**Comment by spudly** [comment]
@novemberborn,

That dependency was added by @jamestalmage in the `babel-plugin-for-integration-tests` branch, which is merged into my code.

**Comment by vdemedes** [comment]
I guess "outside" commits shouldn't be in this PR, should they?

**Comment by novemberborn** [comment]
> That dependency was added by @jamestalmage in the babel-plugin-for-integration-tests branch, which is merged into my code.

Ah!

@vdemedes are you in Gitter at the moment? Let's see about landing the integration test plugin in `master` etc.

**Comment by novemberborn** [comment]
Landed in https://github.com/sindresorhus/ava/commit/a03f8264edbd99d9c6a86827106ad40dfdc42b10 yay! (Messed up the commit so it didn't close automatically… oops)

**Comment by novemberborn** [comment]
Awesome work, thanks @spudly!

**Comment by spudly** [comment]
Nice! Thanks for the help everyone!

**Comment by sindresorhus** [comment]
Yay! Nice work @spudly :dancer: 

od3vvoq

**Comment by kentcdodds** [comment]
Hooray! 🎉🎉🎉

**Comment by Wildhoney** [comment]
:clap: :clap:

For those wondering, it appears as though this hasn't *yet* been published to npm &mdash; however a `npm i https://github.com/sindresorhus/ava` will suffice for the moment.

**Comment by vdemedes** [comment]
giphy 7

**Comment by tleunen** [comment]
Can we get a cli option for that as well? I prefer specifying it there instead of adding a config in the package.json

**Comment by jamestalmage** [comment]
@tleunen

Please raise an issue.
